In [1]:
import pandas as pd
import numpy as np
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report

Given the following confusion matrix, 
evaluate (by hand) the model's performance.

|               | pred dog   | pred cat   |
|:------------  |-----------:|-----------:|
| actual dog    |         46 |         7  |
| actual cat    |         13 |         34 |

In the context of this problem, what is a false positive?
True positive is predicting a cat and finding a cat

In the context of this problem, what is a true negative?
True negative is predicting a cat and not getting a cat

In [34]:
TP = 34
FP = 7
FN = 13
TN = 46

In [35]:
print("Cat-Finder")
print("True Positives", TP)
print("False Positives", FP)
print("False Negatives", FN)
print("True Negatives", TN)

accuracy = (TP + TN) / (TP + TN + FP + FN)
recall = TP / (TP + FN)
precision = TP / (TP + FP)

print("Accuracy is", accuracy)
print("Recall is", round(recall,2))
print("Precision is", round(precision,2))

Cat-Finder
True Positives 34
False Positives 7
False Negatives 13
True Negatives 46
Accuracy is 0.8
Recall is 0.72
Precision is 0.83


You are working as a datascientist working for Codeup Cody Creator 
(C3 for short), a rubber-duck manufacturing plant.

Unfortunately, some of the rubber ducks that are produced will have defects.
Your team has built several models that try to predict those defects, and 
the data from their predictions can be found here.

Use the predictions dataset and 
pandas to help answer the following questions:
    

In [31]:
df = pd.read_csv('https://ds.codeup.com/data/c3.csv')
df.head()

actual     model1  model2     model3
0  No Defect  No Defect  Defect  No Defect
1  No Defect  No Defect  Defect     Defect
2  No Defect  No Defect  Defect  No Defect
3  No Defect     Defect  Defect     Defect
4  No Defect  No Defect  Defect  No Defect

An internal team wants to investigate the cause of the manufacturing defects.
They tell you that they want to identify as many of the ducks that have a defect as possible. 
Which evaluation metric would be appropriate here? 
Which model would be the best fit for this use case?

In [37]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   actual  200 non-null    object
 1   model1  200 non-null    object
 2   model2  200 non-null    object
 3   model3  200 non-null    object
dtypes: object(4)
memory usage: 6.4+ KB


In [38]:
df.actual.value_counts()

No Defect    184
Defect        16
Name: actual, dtype: int64

In [39]:
df["baseline"] = df.actual.value_counts().idxmax()
df.head()

actual     model1  model2     model3   baseline
0  No Defect  No Defect  Defect  No Defect  No Defect
1  No Defect  No Defect  Defect     Defect  No Defect
2  No Defect  No Defect  Defect  No Defect  No Defect
3  No Defect     Defect  Defect     Defect  No Defect
4  No Defect  No Defect  Defect  No Defect  No Defect

In [52]:
classes = ['Defect', 'No Defect']
print(classification_report(df.actual, df.model1, target_names=classes))

              precision    recall  f1-score   support

      Defect       0.69      0.82      0.75      1746
   No Defect       0.89      0.80      0.84      3254

    accuracy                           0.81      5000
   macro avg       0.79      0.81      0.80      5000
weighted avg       0.82      0.81      0.81      5000



In [53]:
print(classification_report(df.actual, df.model2, target_names=classes))

              precision    recall  f1-score   support

      Defect       0.48      0.89      0.63      1746
   No Defect       0.89      0.49      0.63      3254

    accuracy                           0.63      5000
   macro avg       0.69      0.69      0.63      5000
weighted avg       0.75      0.63      0.63      5000



In [54]:
print(classification_report(df.actual, df.model3, target_names=classes))

              precision    recall  f1-score   support

      Defect       0.36      0.51      0.42      1746
   No Defect       0.66      0.51      0.57      3254

    accuracy                           0.51      5000
   macro avg       0.51      0.51      0.50      5000
weighted avg       0.55      0.51      0.52      5000



Recently several stories in the local news have come out highlighting customers 
who received a rubber duck with a defect, and portraying C3 in a bad light. 
The PR team has decided to launch a program that gives customers with 
a defective duck a vacation to Hawaii. 
They need you to predict which ducks will have defects, 
but tell you the really don't want to accidentally give out a vacation package 
when the duck really doesn't have a defect. 
Which evaluation metric would be appropriate here? 
Which model would be the best fit for this use case?

false postive when predict defective but it is not defective
false negative is when we say not defective but it is

calculate precision model1 is the answer

You are working as a data scientist for Gives You Paws ™, a subscription based service that shows you cute pictures of dogs or cats (or both for an additional fee).

At Gives You Paws, anyone can upload pictures of their cats or dogs. The photos are then put through a two step process. First an automated algorithm tags pictures as either a cat or a dog (Phase I). Next, the photos that have been initially identified are put through another round of review, possibly with some human oversight, before being presented to the users (Phase II).

Several models have already been developed with the data, and you can find their results here.

Given this dataset, use pandas to create a baseline model (i.e. a model that just predicts the most common class) and answer the following questions:

In [45]:
url = 'https://ds.codeup.com/data/gives_you_paws.csv'

df = pd.read_csv(url)
df.head()

actual model1 model2 model3 model4
0    cat    cat    dog    cat    dog
1    dog    dog    cat    cat    dog
2    dog    cat    cat    cat    dog
3    dog    dog    dog    cat    dog
4    cat    cat    cat    dog    dog

In [46]:
df['baseline'] = df.actual.value_counts().idxmax()

In [47]:
df.baseline.value_counts()

dog    5000
Name: baseline, dtype: int64

In [48]:
df.actual.value_counts()

dog    3254
cat    1746
Name: actual, dtype: int64

In [49]:
df.actual.value_counts(normalize=True)

dog    0.6508
cat    0.3492
Name: actual, dtype: float64

In [55]:
(df.actual == df.baseline).mean()

0.6508

In terms of accuracy, how do the various models compare to the baseline model? Are any of the models better than the baseline?

In [26]:
confusion_matrix(df.actual, df.model1)

classes = ['cat', 'dog']

print(classification_report(df.actual, df.model1, target_names=classes))
      

              precision    recall  f1-score   support

         cat       0.69      0.82      0.75      1746
         dog       0.89      0.80      0.84      3254

    accuracy                           0.81      5000
   macro avg       0.79      0.81      0.80      5000
weighted avg       0.82      0.81      0.81      5000



In [27]:
confusion_matrix(df.actual, df.model2)
print(classification_report(df.actual, df.model2, target_names=classes))

              precision    recall  f1-score   support

         cat       0.48      0.89      0.63      1746
         dog       0.89      0.49      0.63      3254

    accuracy                           0.63      5000
   macro avg       0.69      0.69      0.63      5000
weighted avg       0.75      0.63      0.63      5000



In [30]:
confusion_matrix(df.actual, df.model3)
print(classification_report(df.actual, df.model3, target_names=classes))

              precision    recall  f1-score   support

         cat       0.36      0.51      0.42      1746
         dog       0.66      0.51      0.57      3254

    accuracy                           0.51      5000
   macro avg       0.51      0.51      0.50      5000
weighted avg       0.55      0.51      0.52      5000



In [29]:
confusion_matrix(df.actual, df.model4)
print(classification_report(df.actual, df.model4, target_names=classes))

              precision    recall  f1-score   support

         cat       0.81      0.35      0.48      1746
         dog       0.73      0.96      0.83      3254

    accuracy                           0.74      5000
   macro avg       0.77      0.65      0.66      5000
weighted avg       0.76      0.74      0.71      5000



Suppose you are working on a team that solely deals with dog pictures. Which of these models would you recomend for Phase I? For Phase II?

Phase I: We use the one with the highest recall for dogs so model4 with 0.96

Suppose you are working on a team that solely deals with cat pictures. Which of these models would you recomend for Phase I? For Phase II?

Phase I: We use the one with the highest recall for cats so model2 with 0.89